src/Proxy.hs

```diff
@@ -321,41 +320,131 @@ parseSetCookie' c =
+          case resultValue of
+            -- Invariant error, e.g. converting the GRPC response payload to Haskell datatype.
+            Left err -> do
+              $(logError) $ "Invariant error: " <> Text.pack err
+              return $
+                Left (InvariantError, internalServerError500, InternalError, EMErrorResponse $ Yesod.InternalError "Invariant error.")
+            -- Otherwise the conversion succeeded.
+            Right val -> do
+              $(logOther "Trace") $ "Got these response headers: " <> Text.pack (show hds)
+              -- set cookies in response
+              let scs = parseSetCookie' . snd <$> (filter $ ("set-cookie" ==) . fst) hds
+              mapM_ setCookie scs
+              $(logOther "Trace") $ "Set-cookies headers to be included in yesod response to client: " <> Text.pack (show scs)
+              -- update cookie map
+              cacheSet $! Map.union updatedCookies cookies
+              return $ Right $ k val
+        -- GRPC response with an invalid status code.
+        StatusInvalid -> do
+          $(logError) "Got invalid GRPC status code."
+          return $
+            Left (StatusInvalidError, badGateway502, InternalError, EMGRPCErrorResponse "Invalid GRPC status code.")
+        -- GRPC response with status code 'NOT_FOUND'.
+        StatusNotOk (NOT_FOUND, err) -> do
+          return $
+            Left (StatusNotOkError NOT_FOUND, notFound404, DataNotFound, EMGRPCErrorResponse $ "Requested object was not found: " <> err)
+        -- GRPC response with valid non-'OK' status code.
+        StatusNotOk (status, err) -> do
+          $(logError) $ "Got non-OK GRPC status code '" <> Text.pack (show status) <> "': " <> Text.pack err
+          return $
+            Left (StatusNotOkError status, badGateway502, InternalError, EMGRPCErrorResponse $ "Non-OK GRPC status code '" <> show status <> "'.")
+        -- GRPC request failed in some other way.
+        RequestFailed err -> do
+          $(logError) $ "GRPC call failed: " <> Text.pack err
+          return $
+            Left (RequestFailedError, badGateway502, InternalError, EMGRPCErrorResponse "Unable to communicate with the node.")
+  case responseInfoOrErrorData of
+    -- An error occurred. Send an error response, potentially overriding default values.
+    Left (errType, status, errCode, errMsg) -> do
+      (status', errCode', errMsg') <- case resp of
+        -- Custom error information was provided.
+        Just (onErrType, statusM, errCodeM, errMsgM) ->
+          -- If the error type matches that of the provided, override with the provided error information.
+          if errType == onErrType
+            then return (fromMaybe status statusM, fromMaybe errCode errCodeM, fromMaybe errMsg errMsgM)
+            else return (status, errCode, errMsg)
+        -- Otherwise, no custom error information was provided.
+        Nothing -> return (status, errCode, errMsg)
```

Comment:
```suggestion
    Left (errType, status, errCode, errMsg) -> do
      let (status', errCode', errMsg') <- case resp of
        -- Custom error information was provided.
        Just (onErrType, statusM, errCodeM, errMsgM) ->
          -- If the error type matches that of the provided, override with the provided error information.
          if errType == onErrType
            then (fromMaybe status statusM, fromMaybe errCode errCodeM, fromMaybe errMsg errMsgM)
            else (status, errCode, errMsg)
        -- Otherwise, no custom error information was provided.
        Nothing -> (status, errCode, errMsg)
```